# Pandas passo a passo — Loja de roupas

Cada linha representa uma peça de roupa. Vamos analisar preços em reais e quantidades disponíveis no estoque.

## Como usar

1. Na primeira parte, tente resolver sem olhar o gabarito.
2. Escreva o código nas células de código e suas explicações nas células de texto.
3. Execute as células de cima para baixo.

**Como pensar:** qual informação preciso? Qual coluna contém essa informação? Quero todas as linhas ou apenas algumas? Qual operação responde à pergunta?

Comentários iniciados por `#` não são executados. `NameError` pode indicar uma preparação não executada; `KeyError` pode indicar um nome de coluna escrito incorretamente.

## Preparação

`import pandas as pd` disponibiliza a biblioteca com o apelido `pd`. `dados` é um dicionário: cada chave vira uma coluna e cada lista fornece seus valores. `pd.DataFrame(dados)` cria a tabela; `df =` guarda essa tabela na variável `df`.

Os números à esquerda são os índices das linhas. Eles começam em zero.

In [1]:
import pandas as pd

dados = {'peca': ['Camiseta', 'Calça', 'Vestido', 'Jaqueta', 'Shorts', 'Saia'], 'preco': [40, 120, 150, 200, 60, 90], 'estoque': [20, 8, 5, 3, 15, 7]}

df = pd.DataFrame(dados)
df

,peca,preco,estoque
0,Camiseta,40,20
1,Calça,120,8
2,Vestido,150,5
3,Jaqueta,200,3
4,Shorts,60,15
5,Saia,90,7


## Parte 1 — perguntas

Para cada questão: **preveja → escreva → execute → interprete**.

### Questão 1

Mostre as três primeiras linhas.

**Pense antes:** Quais peças aparecerão? Isso remove as outras?

**Minha previsão:**

Escreva aqui.

In [2]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 2

Mostre somente os nomes das peças.

**Pense antes:** Qual coluna precisamos selecionar?

**Minha previsão:**

Escreva aqui.

In [3]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 3

Calcule a média dos preços.

**Pense antes:** Qual é a soma dos preços? Por quantas peças devemos dividir?

**Minha previsão:**

Escreva aqui.

In [4]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 4

Mostre as peças que custam menos de R$ 100, com todas as colunas.

**Pense antes:** Uma peça de exatamente 100 entraria?

**Minha previsão:**

Escreva aqui.

In [5]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 5

Use loc para mostrar somente os nomes das peças com estoque menor ou igual a 5.

**Pense antes:** Estoque igual a 5 entra? Qual parte escolhe linhas e qual escolhe coluna?

**Minha previsão:**

Escreva aqui.

In [6]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 6

Use iloc para mostrar as duas primeiras linhas e as duas primeiras colunas.

**Pense antes:** Quais posições entram em 0:2?

**Minha previsão:**

Escreva aqui.

In [7]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 7

Mostre os estoques de todas as peças usando loc.

**Pense antes:** Como representar todas as linhas?

**Minha previsão:**

Escreva aqui.

In [8]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 8

Ordene do maior preço para o menor.

**Pense antes:** Qual peça será a primeira e qual será a última?

**Minha previsão:**

Escreva aqui.

In [9]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 9

Crie uma cópia e altere o preço da Camiseta para R$ 45.

**Pense antes:** Como localizar Camiseta pelo nome? Qual a diferença entre == e =?

**Minha previsão:**

Escreva aqui.

In [10]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 10

Desafio: calcule o valor do estoque de cada peça em uma nova coluna de uma cópia.

**Pense antes:** Se há 20 camisetas de R$ 40, quanto vale esse estoque?

**Minha previsão:**

Escreva aqui.

In [11]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

## Parte 2 — resolução comentada

Compare com suas tentativas. Leia como pensar, execute o código e explique o resultado com suas palavras. Comece executando a preparação novamente.

In [12]:
import pandas as pd

dados = {'peca': ['Camiseta', 'Calça', 'Vestido', 'Jaqueta', 'Shorts', 'Saia'], 'preco': [40, 120, 150, 200, 60, 90], 'estoque': [20, 8, 5, 3, 15, 7]}

df = pd.DataFrame(dados)
df

,peca,preco,estoque
0,Camiseta,40,20
1,Calça,120,8
2,Vestido,150,5
3,Jaqueta,200,3
4,Shorts,60,15
5,Saia,90,7


### Resolução 1

**Pedido:** Mostre as três primeiras linhas.

**Como pensar:** Queremos ver o começo da tabela; usamos head(3).

In [13]:
df.head(3)

,peca,preco,estoque
0,Camiseta,40,20
1,Calça,120,8
2,Vestido,150,5


**Resultado explicado:** Camiseta, Calça e Vestido aparecem. As demais peças continuam na tabela.

### Resolução 2

**Pedido:** Mostre somente os nomes das peças.

**Como pensar:** Selecionamos uma informação de todas as linhas usando o nome da coluna.

In [14]:
df["peca"]

0    Camiseta
1       Calça
2     Vestido
3     Jaqueta
4      Shorts
5        Saia
Name: peca, dtype: object

**Resultado explicado:** Aparecem os seis nomes, sem os preços e estoques.

### Resolução 3

**Pedido:** Calcule a média dos preços.

**Como pensar:** Selecionamos preco e aplicamos mean().

In [15]:
media = df["preco"].mean()
print(media)

110.0


**Resultado explicado:** A soma é 660; 660 / 6 = 110. A média é R$ 110,00, embora nenhuma peça tenha exatamente esse preço.

### Resolução 4

**Pedido:** Mostre as peças que custam menos de R$ 100, com todas as colunas.

**Como pensar:** A condição compara cada preço com 100; os colchetes externos selecionam as linhas verdadeiras.

In [16]:
df[df["preco"] < 100]

,peca,preco,estoque
0,Camiseta,40,20
4,Shorts,60,15
5,Saia,90,7


**Resultado explicado:** Camiseta, Shorts e Saia aparecem. O valor 100 não seria incluído.

### Resolução 5

**Pedido:** Use loc para mostrar somente os nomes das peças com estoque menor ou igual a 5.

**Como pensar:** Antes da vírgula usamos a condição; depois escolhemos peca. <= inclui o limite.

In [17]:
df.loc[df["estoque"] <= 5, "peca"]

2    Vestido
3    Jaqueta
Name: peca, dtype: object

**Resultado explicado:** Vestido e Jaqueta aparecem. Vestido entra porque tem estoque 5.

### Resolução 6

**Pedido:** Use iloc para mostrar as duas primeiras linhas e as duas primeiras colunas.

**Como pensar:** As posições começam em 0. O final 2 fica de fora.

In [18]:
df.iloc[0:2, 0:2]

,peca,preco
0,Camiseta,40
1,Calça,120


**Resultado explicado:** Camiseta e Calça aparecem nas colunas peca e preco.

### Resolução 7

**Pedido:** Mostre os estoques de todas as peças usando loc.

**Como pensar:** O : escolhe todas as linhas; estoque escolhe a coluna.

In [19]:
df.loc[:, "estoque"]

0    20
1     8
2     5
3     3
4    15
5     7
Name: estoque, dtype: int64

**Resultado explicado:** Os valores são 20, 8, 5, 3, 15 e 7.

### Resolução 8

**Pedido:** Ordene do maior preço para o menor.

**Como pensar:** sort_values ordena a coluna. ascending=False pede ordem decrescente.

In [20]:
df.sort_values("preco", ascending=False)

,peca,preco,estoque
3,Jaqueta,200,3
2,Vestido,150,5
1,Calça,120,8
5,Saia,90,7
4,Shorts,60,15
0,Camiseta,40,20


**Resultado explicado:** Jaqueta, Vestido, Calça, Saia, Shorts e Camiseta. O resultado não substitui df.

### Resolução 9

**Pedido:** Crie uma cópia e altere o preço da Camiseta para R$ 45.

**Como pensar:** copy cria a cópia. == identifica a peça; = atribui o novo preço à coluna selecionada.

In [21]:
df_alterado = df.copy()
df_alterado.loc[df_alterado["peca"] == "Camiseta", "preco"] = 45
df_alterado

,peca,preco,estoque
0,Camiseta,45,20
1,Calça,120,8
2,Vestido,150,5
3,Jaqueta,200,3
4,Shorts,60,15
5,Saia,90,7


**Resultado explicado:** A cópia tem Camiseta a 45; a original continua com 40.

### Resolução 10

**Pedido:** Desafio: calcule o valor do estoque de cada peça em uma nova coluna de uma cópia.

**Como pensar:** Multiplicamos preço por quantidade, linha a linha, e guardamos em uma nova coluna.

In [22]:
df_valores = df.copy()
df_valores["valor_estoque"] = df_valores["preco"] * df_valores["estoque"]
df_valores

,peca,preco,estoque,valor_estoque
0,Camiseta,40,20,800
1,Calça,120,8,960
2,Vestido,150,5,750
3,Jaqueta,200,3,600
4,Shorts,60,15,900
5,Saia,90,7,630


**Resultado explicado:** Valores: 800, 960, 750, 600, 900 e 630 reais. Isso é valor de estoque ao preço de venda, não lucro.

## Lembretes

| Recurso | Significado |
|---|---|
| `df["coluna"]` | Seleciona uma coluna |
| `df[condicao]` | Seleciona as linhas que atendem à condição |
| `df.loc[linhas, colunas]` | Seleciona por rótulos ou condições |
| `df.iloc[linhas, colunas]` | Seleciona por posições começando em zero |
| `:` sozinho | Todas as linhas ou colunas |
| `0:3` no iloc | Posições 0, 1 e 2; exclui a posição 3 |
| `==` | Compara valores |
| `=` | Atribui um valor |

Selecionar ou mostrar um resultado não significa apagar as outras informações. A alteração de valores é feita explicitamente, como na questão 9.

Documentação: https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html